### ISSUE TITLE
503 Service Unavailable error when requesting 'admin/fastlyCdn_Edge_Dictionary/ListAll'

### ISSUE THREAD

**Issue by hostep** [fastly/fastly-magento2#163]
Hi guys

Yesterday we enabled Fastly on a staging environment of the Magento Cloud.

But we sometimes see errors popping up in random adminhtml pages under Stores > Configuration (not only in Advanced > System), this error is displayed in a red box: `A technical problem with the server created an error. Try again to continue what you were doing. If the problem persists, try again later.`

When looking through the Network tab of my inspector, I can see that some XHR requests to the Fastly module throws a 503 error which causes this problem.

I can see 3 similar XHR requests, two of them work and one fails.
The ones which work:
- https://example.com/my-super-secret/admin/fastlyCdn_Vcl/getBackends/key/aabbccddeeff/?active_version=8
- https://example.com/my-super-secret/admin/fastlyCdn_Edge_Acl/ListAll/key/aabbccddeeff/?active_version=8

The one which fails with a 503 error:
- https://example.com/my-super-secret/admin/fastlyCdn_Edge_Dictionary/ListAll/key/aabbccddeeff/?active_version=8

And I also sometimes see this one fail with a 503 error:
- https://example.com/my-super-secret/admin/fastlyCdn_Vcl/checkAuthDictionary/key/aabbccddeefff/?isAjax=true

It's a bit random these errors, sometimes it's the first one, sometimes the other one, sometimes even both, and then sometimes no errors, really strange behavior.

Looking at the actual contents of the response, we see something like this (it feels like this is a server error somehow, but no idea how to debug this):
```html
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html>
  <head>
    <title>503 Service Unavailable</title>
  </head>
  <body>
    <h1>Error 503 Service Unavailable</h1>
    <p>Service Unavailable</p>
    <h3>Guru Mediation:</h3>
    <p>Details: cache-ams4432-AMS some-number some-other-number</p>
    <hr>
    <p>Varnish cache server</p>
  </body>
</html>
```

Extra info which might be relevant:
- we have no Edge Dictionaries setup at the moment.
- Magento 2.2.2
- `fastly/magento2` version 1.2.48

Any ideas what can cause this or how we can solve this?

If there are no errors, those request return valid looking json strings, so I don't think the problem is in the Fastly module itself, but actually on the server side. If we have to report this to the Magento Cloud support instead, please let us know.

Thanks!

**Comment by vvuksan** [comment]
We are looking into it.

**Comment by tnikcevs** [comment]
Hello

I've attempted to reproduce the issue on Magento versions (2.2.2 and 2.2.3) and both Fastly versions 1.2.48 and 1.2.49 (last) and was unable to reproduce the issue. I've used clean version with only domain set (without any config) and our working version (with old config) in testing.

If possible please check logs for server errors and let us know if this is something related to part of the fastly code so we will look into it.

also if you used active_version=8, can you let us know what config you have adjusted in version 8 (have you used magento admin to update config or did you make custom adjustments?)

Regards,
Tomislav

**Comment by jordanvector** [comment]
I can confirm that this is an issue happening in a magneto-cloud pro environment.  I am seeing this error on this call on every single page load in the admin:

/admin/admin/fastlyCdn_Vcl/getBackends/key/6a6c898626bdabaec07b2e044c54e7c2e9b309e1b44df8151a81dceff3b19ba8/

Exception: Noticed exception 'Exception' with message 'Warning: session_start(): Failed to read session data: user (path: /var/lib/php/sessions) in /app/7alebf6ygqtlm/vendor/magento/framework/Session/SessionManager.php on line 189' in /app/7alebf6ygqtlm/vendor/magento/framework/App/ErrorHandler.php:61
in Magento\Framework\App\ErrorHandler::handler called at ? (?)
in session_start called at /app/7alebf6ygqtlm/vendor/magento/framework/Session/SessionManager.php (189)
in Magento\Framework\Session\SessionManager::start called at /app/7alebf6ygqtlm/generated/code/Magento/Backend/Model/Session/Interceptor.php (63)
in Magento\Backend\Model\Session\Interceptor::start called at /app/7alebf6ygqtlm/vendor/magento/framework/Session/SessionManager.php (130)
in Magento\Framework\Session\SessionManager::__construct called at /app/7alebf6ygqtlm/generated/code/Magento/Backend/Model/Session/Interceptor.php (14)
in Magento\Backend\Model\Session\Interceptor::__construct called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (111)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::createObject called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (66)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (144)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgument called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (230)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (34)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::_resolveArguments called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (59)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (144)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgument called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (230)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (34)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::_resolveArguments called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (59)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (144)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgument called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (230)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (34)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::_resolveArguments called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (59)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (144)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgument called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (230)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (34)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::_resolveArguments called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (59)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (144)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgument called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (230)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (34)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::_resolveArguments called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (59)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (144)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgument called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (230)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (34)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::_resolveArguments called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (59)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (144)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgument called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php (230)
in Magento\Framework\ObjectManager\Factory\AbstractFactory::resolveArgumentsInRuntime called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (34)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::_resolveArguments called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php (59)
in Magento\Framework\ObjectManager\Factory\Dynamic\Developer::create called at /app/7alebf6ygqtlm/vendor/magento/framework/ObjectManager/ObjectManager.php (70)
in Magento\Framework\ObjectManager\ObjectManager::get called at /app/7alebf6ygqtlm/vendor/magento/framework/Interception/PluginList/PluginList.php (250)
in Magento\Framework\Interception\PluginList\PluginList::getPlugin called at /app/7alebf6ygqtlm/vendor/magento/framework/Interception/Interceptor.php (119)
in Magento\Framework\App\FrontController\Interceptor::Magento\Framework\Interception\{closure} called at /app/7alebf6ygqtlm/vendor/magento/framework/Interception/Interceptor.php (153)
in Magento\Framework\App\FrontController\Interceptor::___callPlugins called at /app/7alebf6ygqtlm/generated/code/Magento/Framework/App/FrontController/Interceptor.php (26)
in Magento\Framework\App\FrontController\Interceptor::dispatch called at /app/7alebf6ygqtlm/vendor/magento/framework/App/Http.php (135)
in Magento\Framework\App\Http::launch called at /app/7alebf6ygqtlm/vendor/magento/framework/Interception/Interceptor.php (58)
in Magento\Framework\App\Http\Interceptor::___callParent called at /app/7alebf6ygqtlm/vendor/magento/framework/Interception/Interceptor.php (138)
in Magento\Framework\App\Http\Interceptor::Magento\Framework\Interception\{closure} called at /app/7alebf6ygqtlm/vendor/msp/shield/Plugin/AppInterfacePlugin.php (111)
in MSP\Shield\Plugin\AppInterfacePlugin::aroundLaunch called at /app/7alebf6ygqtlm/vendor/magento/framework/Interception/Interceptor.php (135)
in Magento\Framework\App\Http\Interceptor::Magento\Framework\Interception\{closure} called at /app/7alebf6ygqtlm/vendor/magento/framework/Interception/Interceptor.php (153)
in Magento\Framework\App\Http\Interceptor::___callPlugins called at /app/7alebf6ygqtlm/generated/code/Magento/Framework/App/Http/Interceptor.php (26)
in Magento\Framework\App\Http\Interceptor::launch called at /app/7alebf6ygqtlm/vendor/magento/framework/App/Bootstrap.php (256)
in Magento\Framework\App\Bootstrap::run called at /app/7alebf6ygqtlm/pub/index.php (37)



I am also not sure why its looking for sessions in var/lib/php/sessions.  Also I am on the latest fastly extension 1.2.49.  Thoughts?

**Comment by vvuksan** [comment]
@jordanvector what happens if you revert to an older version ?

**Comment by jordanvector** [comment]
@vvuksan good question, I can't remember where started but I am pretty sure it was with the last 2 versions.  I can revert 2 versions and report back

**Comment by vvuksan** [comment]
That would be great @jordanvector as we are not able to reproduce this.

**Comment by jordanvector** [comment]
@vvuksan I have noticed this typically after a deployment on magento cloud just fyi

**Comment by vvuksan** [comment]
Does the issue clear itself or does it persist ?

**Comment by jordanvector** [comment]
the issue persists but I was able to get rid of it for a period of time by adding a dictionary in the backend options

**Comment by vvuksan** [comment]
That's helpful. Just to confirm to workaround the issue all you had to do was Edge Dictionary ?

**Comment by jordanvector** [comment]
@vvuksan correct the workaround for me was to add an edge dictionary

**Comment by jordanvector** [comment]
@udovicic so we are definitely using redis for session storage which is very odd that it is looking at /var/lib/php/sessions. Here is the quick line from env.php (minus connection values)

  'session' => 
  array (
    'save' => 'redis',
    'redis' =>

**Comment by jordanvector** [comment]
@udovicic upon further investigation sessions are be written to redis (so not quite sure whats going on there).  I have a ticket open with magento as well, will update if I hear anything from them

**Comment by jordanvector** [comment]
@udovicic after digging into this it seems like my issue specifically was not linked to fastly.   the magento cloud php.ini by default has sessions configured to save to the filesystem by default which was causing the error that I was seeing.  When I updated the php.ini to save to redis like this:

**Comment by yyevgenii** [comment]
The issue is caused by Redis session blocking mechanism. The Cm/RedisSession library has a variable which limits the maximum number of processes that can wait for lock on one session. By default it's value is 6. The default value leads to errors on Magento frontend/backend when, for example, a web page makes several AJAX requests to Magento.

That's what you see when navigate to Stores > Configuration > System: Fastly extension starts sending several AJAX requests to Magento and you reach the maximum concurrency limit. When it happens, Magento responds with 503 error and you can find "Area code is not set" errors in the var/report/ directory.

To avoid the issue, you have to increase the max_concurrency value in app/etc/env.php:
```
'session' =>
        array (
            'save' => 'redis',
            'redis' =>
                array (
                    'host' => '127.0.0.1',
                    'port' => '6379',
                    'database' => 0,
                    'max_concurrency' => 20,
                ),
        ),
```

If you're a Magento Cloud user, you mustn't edit env.php manually, but add the SESSION_CONFIGURATION variable with Redis configuration in JSON format. You can get more details by submitting a support request.

**Comment by yyevgenii** [comment]
@vvuksan it's not a fastly/magtento2 bug, you can close the issue.

**Comment by vvuksan** [comment]
wow thanks :-)

**Comment by hostep** [comment]
Thanks for the info @yyevgenii!

I'll test this when I find some time to work on this project again (probably next week).

**Comment by agorbulin** [comment]
I confirmed, that after changing session settings issue was fixed.
For Magento cloud you can just make env variable SESSION_CONFIGURATION (type is json) with content:
{
  "save": "redis",
  "redis": {
    "host": "localhost",
    "port": "6372",
    "database": 0,
    "max_concurrency": 20
  }
}
**Note:** your environment settings can be differ

**Comment by hostep** [comment]
Can confirm the fix from @agorbulin.

I first tried to change the concurrency value using the `.magento.env.yaml` file for the Magento Cloud hosting, but it's like @agorbulin is saying: some settings (host, port) are different over different environments, so we had to also define that environment variable using the Cloud UI per environment specific.

And it solves the problem indeed!

**Comment by dborishansky** [comment]
Increasing `max_concurrency` did address the errors in the admin / logs, but most unfortunately introduced performance issues on the front end. Still hunting for a viable solution to this issue. 

The burden of responsibility here seems to fall to the fastly module, if reducing the number of AJAX requests is feasible?

**Comment by vvuksan** [comment]
Can you elaborate on performance issues on the front end ? I'm a bit surprised that one cannot make more than 6 AJAX requests.

**Comment by vvuksan** [comment]
Sounds like a lock contention.

**Comment by vvuksan** [comment]
We'll look into doing this differently.

**Comment by alanablett** [comment]
Getting the same issue here on Cloud. Using those redis settings caused users to be logged out intermittently so that's a no go for us unfortunately.

**Comment by vvuksan** [comment]
We have addressed this in #177. Fixes were released as part of 1.2.54.

**Comment by hostep** [comment]
Just tested by removing the `SESSION_CONFIGURATION` variable again and upgrading to version 1.2.54 and all looks fixed indeed. Thanks @vvuksan!
